test/hooks.js

```diff
@@ -245,3 +246,49 @@ test('`afterResponse` hook gets called even if using body shortcuts', async t =>
 
 	await server.close();
 });
+
+test('beforeRetry hook is never called for the initial request', async t => {
+	const server = await createTestServer();
+	server.get('/', async (request, response) => {
+		response.end(request.headers.unicorn);
+	});
+
+	t.not(await ky.get(server.url, {
+		hooks: {
+			beforeRetry: [
+				options => {
+					options.headers.set('unicorn', fixture);
+				}
+			]
+		}
+	}).text(), fixture);
+
+	await server.close();
+});
+
+test('beforeRetry hook allows modifications of non initial requests', async t => {
+	let requestCount = 0;
+
+	const server = await createTestServer();
+	server.get('/', async (request, response) => {
+		requestCount++;
+
+		if (requestCount > 1) {
+			response.end(request.headers.unicorn);
+		} else {
+			response.sendStatus(408);
+		}
+	});
+
+	t.is(await ky.get(server.url, {
+		hooks: {
+			beforeRetry: [
+				options => {
+					options.headers.set('unicorn', fixture);
+				}
+			]
+		}
+	}).text(), fixture);
+
+	await server.close();
+});
```

Comment:
Make sure you test as much as done for `beforeRetry` in Got: https://github.com/sindresorhus/got/blob/master/test/hooks.ts